pkg/resource/manager.go

```diff
@@ -125,12 +144,245 @@ func (m *manager) BatchCreateGenericResources(ctx context.Context, request *pb.B
+
+	// build a list of generic resource versions from occurrences. these may or may not already exist
+	for _, occurrence := range occurrences {
+		newVersions := genericResourceVersionsFromOccurrence(occurrence)
+
+		// check if we already know about these versions
+		for versionId, newVersion := range newVersions {
+			if existingVersion, ok := genericResourceVersions[versionId]; ok {
+				// if we do, update the names and create time, if needed
+				if newVersion.Created != nil {
+					existingVersion.Created = newVersion.Created
+				}
+
+				if newVersion.Names != nil {
+					existingVersion.Names = newVersion.Names
+				}
+			} else {
+				genericResourceVersions[versionId] = newVersion
+				versionIds = append(versionIds, versionId)
+			}
+		}
+	}
+
+	// check which generic resource versions already exist
+	multiGetResponse, err := m.esClient.MultiGet(ctx, &esutil.MultiGetRequest{
+		DocumentIds: versionIds,
+		Index:       m.indexManager.AliasName(genericResourcesDocumentKind, ""),
+	})
+	if err != nil {
+		return err
+	}
+
+	// versions that don't exist need to be created
+	// versions that do exist may need to be updated
+	var bulkItems []*esutil.BulkRequestItem
+	for i, versionId := range versionIds {
+		version := genericResourceVersions[versionId]
+		uriParts, err := parseResourceUri(versionId)
+		if err != nil {
+			return err
+		}
+
+		bulkItem := &esutil.BulkRequestItem{
+			Operation:  esutil.BULK_INDEX,
+			Message:    version,
+			DocumentId: versionId,
+			Join: &esutil.EsJoin{
+				Field:  genericResourceDocumentJoinField,
+				Name:   genericResourceVersionRelationName,
+				Parent: uriParts.prefixedName,
+			},
+		}
+
+		if multiGetResponse.Docs[i].Found {
+			// if the version already exists, we may have to update(index) it.
+			// if we have a list of names, update the existing version with the new one
+			// otherwise, we have nothing to do here
+			if len(version.Names) != 0 {
+				bulkItem.Operation = esutil.BULK_INDEX
+				log.Debug("updating generic resource version", zap.Any("version", version))
+			} else {
+				bulkItem = nil
+			}
+		} else {
+			// if the version doesn't exist, we need to create it
+			// before creating the version, add a timestamp if it doesn't exist
+			if version.Created == nil {
+				version.Created = timestamppb.Now()
+			}
+
+			bulkItem.Operation = esutil.BULK_CREATE
+			log.Debug("creating generic resource version", zap.Any("version", version))
+		}
+
+		if bulkItem != nil {
+			bulkItems = append(bulkItems, bulkItem)
+		}
+	}
+
+	// perform bulk create / update
+	var bulkItemErrors []error
+	if len(bulkItems) != 0 {
+		bulkResponse, err := m.esClient.Bulk(ctx, &esutil.BulkRequest{
+			Index:   m.indexManager.AliasName(genericResourcesDocumentKind, ""),
+			Refresh: m.esConfig.Refresh.String(),
+			Items:   bulkItems,
+		})
+		if err != nil {
+			return err
+		}
+
+		for _, item := range bulkResponse.Items {
+			var itemError error
+			if item.Create != nil && item.Create.Error != nil {
```

Comment:
Concurrent `BatchCreateOccurrences` could potentially cause us to try to create the same version twice, we may want to ignore the error if the type is `resource_already_exists_exception`

Reply:
Yeah, I wasn't 100% sure what to do about this.  I could also make this an `Index` instead of a `Create` operation, so it will just upsert.  What do you think about that?

Reply:
That sounds good to me.